pair/src/contract.rs

```diff
@@ -0,0 +1,760 @@
+        if recipient == self.token.0 || recipient == self.token.1 {
+            Err(Error::InvalidRecipient)
+        } else {
+            Ok(())
+        }
+    }
+
+    async fn swap_tokens_for_exact_tokens(
+        &mut self,
+        (tx_manager, action_kind): (&mut TransactionManager<CachedAction>, ActionKind),
+        msg_source: ActorId,
+        out_amount: u128,
+        max_in_amount: u128,
+        to: ActorId,
+        swap_kind: SwapKind,
+    ) -> Result<Event, Error> {
+        self.check_recipient(to)?;
+
+        let swap_pattern = self.swap_pattern(swap_kind);
+        let in_amount = calculate_in_amount(out_amount, swap_pattern.reserve)?;
+
+        let mut tx_guard = tx_manager.asquire_transaction(
+            msg_source,
+            action_kind.to_tx_kind(CachedAction::Swap(in_amount)),
+        )?;
+
+        tx_guard
+            .tx_data
+            .check_tx_data(|tx_data| tx_data == &CachedAction::Swap(in_amount))?;
+
+        if in_amount > max_in_amount {
+            return Err(Error::InsufficientFormerAmount);
+        }
+
+        self.swap(
+            &mut tx_guard.stepper,
+            msg_source,
+            swap_kind,
+            (in_amount, out_amount),
+            to,
+            swap_pattern,
+        )
+        .await
+    }
+
+    async fn swap(
+        &mut self,
+        stepper: &mut Stepper,
+        msg_source: ActorId,
+        kind: SwapKind,
+        (in_amount, out_amount): (u128, u128),
+        to: ActorId,
+        SwapPattern {
+            token: (in_token, out_token),
+            reserve,
+            normalize_balance,
+        }: SwapPattern,
+    ) -> Result<Event, Error> {
+        let program_id = exec::program_id();
+
+        utils::transfer_tokens(stepper, in_token, msg_source, program_id, in_amount).await?;
+
+        if let Err(error) =
+            utils::transfer_tokens(stepper, out_token, program_id, to, out_amount).await
+        {
+            utils::transfer_tokens(stepper, in_token, program_id, msg_source, in_amount).await?;
+
+            return Err(error);
+        }
+
+        self.update(normalize_balance((
+            reserve.0 + in_amount,
+            reserve.1 - out_amount,
+        )));
+
+        Ok(Event::Swap {
+            sender: msg_source,
+            in_amount,
+            out_amount,
+            to,
+            kind,
+        })
+    }
+}
+
+struct SwapPattern {
+    token: (ActorId, ActorId),
+    reserve: (u128, u128),
+    normalize_balance: fn((u128, u128)) -> (u128, u128),
+}
+
+fn check_deadline(deadline: u64) -> Result<(), Error> {
+    if exec::block_timestamp() > deadline {
+        Err(Error::DeadlineExceeded)
+    } else {
+        Ok(())
+    }
+}
+
+#[no_mangle]
+extern "C" fn init() {
+    let result = process_init();
+    let is_err = result.is_err();
+
+    reply(result).expect("failed to encode or reply from `init()`");
+
+    if is_err {
+        exec::exit(ActorId::zero());
```

Comment:
Is `exit` better than just `panic` in `init`?

Reply:
Panicking is definitely better, but it'll prevent a contract from replying with a graceful error.